File: 500.600/543.DiameterOfBinaryTree.cpp
#include <iostream>
using namespace std;

struct TreeNode
{
    int val{};
    TreeNode *left{};
    TreeNode *right{};
    TreeNode() {}
    TreeNode(int x) : val(x) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class DiameterOfBinaryTree
{
public:
    int diameterOfBinaryTree(TreeNode *root)
    {
        mlen = 0;
        dfs(root);
        return mlen;
    }

private:
    int dfs(TreeNode *n)
    {
        int l = 0, r = 0;
        if (n->left)
        {
            l = dfs(n->left) + 1;
        }
        if (n->right)
        {
            r = dfs(n->right) + 1;
        }
        mlen = max(mlen, l + r);
        return max(l, r);
    }

private:
    int mlen{};
};

void del(TreeNode *root)
{
    if (root->left)
    {
        del(root->left);
    }
    if (root->right)
    {
        del(root->right);
    }
    delete root;
}

int main()
{
    DiameterOfBinaryTree cls;
    TreeNode *head = new TreeNode(1);
    head->left = new TreeNode(2);
    head->left->left = new TreeNode(4);
    head->left->right = new TreeNode(5);
    head->right = new TreeNode(3);
    cout << cls.diameterOfBinaryTree(head) << " -> 3" << endl;
    del(head);

    head = new TreeNode(1);
    head->left = new TreeNode(2);
    cout << cls.diameterOfBinaryTree(head) << " -> 1" << endl;
    del(head);
    return 0;
}

File: go/500.600/543.DiameterOfBinaryTree.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var mlen = 0

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

func diameterOfBinaryTree(root *TreeNode) int {
	mlen = 0
	dfs(root)
	return mlen
}

func dfs(root *TreeNode) int {
	l, r := 0, 0
	if root.Left != nil {
		l = dfs(root.Left) + 1
	}
	if root.Right != nil {
		r = dfs(root.Right) + 1
	}
	mlen = max(mlen, l+r)
	return max(l, r)
}

func main() {
	head := &TreeNode{Val: 1}
	head.Left = &TreeNode{Val: 2}
	head.Left.Left = &TreeNode{Val: 4}
	head.Left.Right = &TreeNode{Val: 5}
	head.Right = &TreeNode{Val: 3}
	fmt.Println(diameterOfBinaryTree(head), 3)

	head = &TreeNode{Val: 1}
	head.Left = &TreeNode{Val: 2}
	fmt.Println(diameterOfBinaryTree(head), 1)
}
